Fix Vec3 equality and length, and flesh out its tests

hashbangstudio opened the pull request:
As **cmp** no longer in Python 3 need to use functools.total_ordering which is in 2.7 and 3.*
  https://docs.python.org/2/library/functools.html#functools.total_ordering
  https://docs.python.org/3/library/functools.html#functools.total_ordering
  Define **eq** and **lt** and others are created automatically

Add unittest style for file vec3.py

hashbangstudio committed:
Fix for rich comparison and add test for vec3 file

As __cmp__ no longer in Python 3 need to use functools.total_ordering which is in 2.7 and 3.*
  https://docs.python.org/2/library/functools.html#functools.total_ordering
  https://docs.python.org/3/library/functools.html#functools.total_ordering
  Define __eq__ and __lt__ and others are created automatically

Add unittest style for file vec3.py

doismellburning commented:
Thanks for the improved test suite! Would you mind not deleting the existing one though please.

As for ordering, I think conventionally vectors are only partially orderable? (consider `(0,1)` and `(1,0)`) - @pozorvlak probably has better maths-fu than I to confirm that - and `__cmp__` is for total orderings, so shall we (subject to checking that nothing actually needs it) fry it?

hashbangstudio committed:
Add unittest as separate file

hashbangstudio committed:
Revert to original verison of test file

ghickman commented:
@hashbangstudio – Could you drop the addition of `@property` decorators too please? It seems out of scope for this PR.

doismellburning commented on minecraft/vec3.py:
```diff
@@ -58,17 +62,17 @@ def _map(self, func):
+        dx = self.x - rhs.x
+        dy = self.y - rhs.y
+        dz = self.z - rhs.z
+        return (dx < 0 or dy < 0 or dz < 0)
```
This would be a very confusing implementation of `__lt__` because `(0,1) < (1,0) && (1,0) < (0,1)` so it's not even a partial ordering (violates antisymmetry)

I think you want `and` not `or`, c.f. https://en.wikipedia.org/wiki/Ordered_vector_space

hashbangstudio committed:
Remove ordering and property dceorators

Use just __eq__ and __ne__ fro comparison
Remove property decorator
Add missing function brackets in length function
Update test in response

hashbangstudio committed:
Remove runner code

doismellburning commented:
Can someone please explain how `testComparison` passes given `__cmp__` was removed?

doismellburning commented:
Ohhh, because all the ordering tests went away, that'd do it!

hashbangstudio commented:
Do people think it should be renamed `testEquality` or something similar to avoid confusion?

dbrgn commented:
I don't want to be nitpicky, but if you're already renaming, you should be using under_scores for method names (PEP8) :)

ghickman commented:
Seeing as these are new tests could you use the pythonic/pytest style please.

dbrgn commented on minecraft/vec3.py:
```diff
@@ -19,7 +19,7 @@ def __iadd__(self, rhs):
 
     def length(self):
-        return self.lengthSqr ** .5
+        return self.lengthSqr() ** .5
```
This example shows that the code was probably never used much in real life. I'm not sure if we should invest a lot of effort into the Vec3 class.

ghickman replied:
I agree but I'd rather test what's there before we make any decisions about razing it to the ground given the likely usage numbers of the original code base.

doismellburning replied:
++

hashbangstudio commented:
Tests are not wholly new, created from some quickly created tests for original API a while back
CamelCase/mixedCase is a hard habit to break, especially when coding quickly :)
Hence non PEP8 compliance on naming

hashbangstudio committed:
Move to PEP8 naming

Change variables and functions to lower_case

ghickman commented:
@hashbangstudio – converting the tests will be very simple, as an example the `testIteration` method would be converted like so:

```
def test_iteration():
    coords = [1, 9, 6]
    v = Vec3(coords[0], coords[1], coords[2])
    for index, pos in enumerate(v):
        assert pos == coords[index]
```

The changes here were:
- underscore_names
- make a function (not technically necessary)
- switch unittest camelCase methods for asserts with python comparators

hashbangstudio committed:
Delete old unit test

remove unit test to be replaced with py.test style

hashbangstudio committed:
Add more tests and put in class

Put all tests in class so can use marker for class
Add additional missing tests
Move existing tests into the named function

doismellburning commented:
@hashbangstudio Looks good, thanks - in future PRs if you could not change a load of test functionality at the same time as the thing under test please (because you've not only added to the behaviour of `Vec3` you've moved around its previous test suite, which means a somewhat tedious process of verifying that you didn't actually remove any of that previous test suite)

Thanks :)
